### PR TITLE
Upgrade thanos

### LINKED ISSUE #1023: [3] Upgrade Thanos to v0.31.0
**What Chart should be upgraded?**

upstream/thanos, also check and update the Prometheus rules and alerts, additionally enable it to be configured using application credentials with Swift.

https://github.com/bitnami/charts/blob/main/bitnami/thanos/Chart.yaml

**Why should this Chart be upgraded?**

- [ ] We need to perform a security upgrade.
- [ ] We need it for a new feature: *which feature (link to blocked issue)*.
- [x] We need to keep up to prevent future issues.
- [ ] Other: *Write the reason here*

**Acceptance criteria**

- I checked the migration of the new Chart:
    - [ ] I upgraded a Chart and determined that no migration steps are needed.
    - [ ] I upgraded a Chart and added [migration steps](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration).
- [ ] I tested the functionality provided by the new Chart (e.g., Auth flow, Grafana dashboards, etc.)

### PR DESCRIPTION
**Pull request by OlleLarsson**
**What this PR does / why we need it**:
Upgrades thanos helm chart and consequently the thanos application version.

Also adds the ability to use application credentials when using the swift storage backend.

**Which issue this PR fixes**:
fixes #1023 

**Public facing documentation PR** *(if applicable)*

**Special notes for reviewer**:

Didn't add any migration guide for this as I didn't find anything necessary.

The same rules for when updating the object storage config still applies as before, where you need to manually rollout the pods that interacts with object strorage so that they load the latest secret version.

**Add a screenshot or an example to illustrate the proposed solution:**

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [x] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.
- Chart checklist (pick exactly one):
    - [ ] I upgraded no Chart.
    - [ ] I upgraded a Chart and determined that no migration steps are needed.
    - [x] I upgraded a Chart and added [migration steps](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration).

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

### CODE REVIEW

**Comment by crssnd** [comment]
did you check if we need to update the prometheus rules and alerts and grafana dashboards?

**Comment by crssnd** [comment]
in my case after the upgrade receiver distributor went into oom for quite a long time. 
did you have similar issues? maybe we can increase the memory limit

File: helmfile/upstream/thanos/values.yaml [hunk 1]
```diff
@@ -3327,6 +3752,9 @@ receiveDistributor:
   ## @param receiveDistributor.replicaLabel Label to treat as a replica indicator along which data is de-duplicated
   ##
   replicaLabel: replica
+  ## @param receiveDistributor.replicationFactor Thanos Receive Distributor replication-factor
+  ##
+  replicationFactor: 1
```
Comment (aarnq) [inline helmfile/upstream/thanos/values.yaml:3757]: You should change in the release templates so we also configure this here.

Comment (OlleLarsson) [reply helmfile/upstream/thanos/values.yaml:3757]: I think i got what you meant 😅 see [this](https://github.com/elastisys/compliantkubernetes-apps/pull/1600/commits/135c03d11b37b9c8f3752a8386f23dc8852dec41) commit.

**Comment by OlleLarsson** [comment]
> in my case after the upgrade receiver distributor went into oom for quite a long time. did you have similar issues? maybe we can increase the memory limit

I seem to have set the memory limit to 1Gi
```yaml
thanos:
  receiveDistributor:
    resources:
      limits:
        memory: 1000Mi
```
so I would assume that I ran into the same issue as you, but I'll try running an upgrade again to see if it happens again

**Comment by OlleLarsson** [comment]
> so I would assume that I ran into the same issue as you, but I'll try running an upgrade again to see if it happens again

Did not encounter this in my latest test.

File: helmfile/values/thanos/objectstorage-secret.yaml.gotmpl [hunk 2]
```diff
@@ -26,17 +26,13 @@ objstoreConfig: |-
   config:
     auth_version: {{ $swift.authVersion }}
     auth_url: {{ $swift.authUrl }}
-    {{- if getOrNil "region" $swift }}
-    region_name: {{ $swift.region }}
```
Comment (Xartos) [inline helmfile/values/thanos/objectstorage-secret.yaml.gotmpl:30]: Is region not needed any more? Or has it never been?

File: helmfile/values/thanos/objectstorage-secret.yaml.gotmpl [hunk 3]
```diff
@@ -26,17 +26,13 @@ objstoreConfig: |-
...
-    {{- if getOrNil "domainId" $swift }}
-    domain_id: {{ $swift.domainId }}
-    {{- else if getOrNil "domainName" $swift }}
-    domain_name: {{ $swift.domainName }}
-    {{- end }}
+    {{- if getOrNil "applicationCredentialID" $swift }}
+    application_credential_id: {{ $swift.applicationCredentialID }}
+    application_credential_secret: {{ $swift.applicationCredentialSecret }}
+    {{- else }}
     username: {{ $swift.username | quote }}
```
Comment (robinAwallace) [inline helmfile/values/thanos/objectstorage-secret.yaml.gotmpl:36]: Should this be protected if application cred is used instead?

File: helmfile/values/thanos/objectstorage-secret.yaml.gotmpl (hunk 2 above)
Comment (OlleLarsson) [reply helmfile/values/thanos/objectstorage-secret.yaml.gotmpl:30]: It seems to work without, but no I did not intend to remove it (i think :P).

**Review by aarnq — APPROVED** [review]

**Pull request merged**